src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnection.cs

```diff
@@ -50,47 +49,130 @@ protected override Task DisposeConnection()
+        protected override async Task OnDisconnectedAsync(CloseConnectionMessage closeConnectionMessage)
         {
             var connectionId = closeConnectionMessage.ConnectionId;
-            PerformDisconnectCore(connectionId);
+            if (_clientConnections.TryGetValue(connectionId, out var transport))
+            {
+                try
+                {
+                    await transport.MessageChannel.Writer.WriteAsync(closeConnectionMessage);
+                }
+                catch (Exception e)
+                {
+                    Log.FailToWriteMessageToApplication(_logger, connectionId, e);
+                }
+            }
+        }
 
-            return Task.CompletedTask;
+        protected override async Task OnMessageAsync(ConnectionDataMessage connectionDataMessage)
+        {
+            var connectionId = connectionDataMessage.ConnectionId;
+            if (_clientConnections.TryGetValue(connectionId, out var transport))
+            {
+                try
+                {
+                    await transport.MessageChannel.Writer.WriteAsync(connectionDataMessage);
+                }
+                catch (Exception e)
+                {
+                    Log.FailToWriteMessageToApplication(_logger, connectionId, e);
+                }
+            }
         }
 
-        protected override Task OnMessageAsync(ConnectionDataMessage connectionDataMessage)
+        private async Task PerformDisconnectCore(string connectionId)
         {
-            if (_clientConnections.TryGetValue(connectionDataMessage.ConnectionId, out var transport))
+            if (_clientConnections.TryRemove(connectionId, out var transport))
+            {
+                try
+                {
+                    // Mark channel complete and wait for reader
+                    transport.MessageChannel.Writer.TryComplete();
+                    await transport.MessageChannel.Reader.Completion;
+                }
+                catch (Exception e)
+                {
+                    Log.ApplicaitonTaskFailed(_logger, e);
+                }
+                finally
+                {
+                    transport.OnDisconnected();
+                }
+            }
+
+            Log.ConnectedEnding(_logger, connectionId);
+        }
+
+        private async Task OnConnectedAsyncCore(OpenConnectionMessage message)
+        {
+            var connectionId = message.ConnectionId;
+            if (_clientConnections.TryGetValue(connectionId, out var transport))
+            {
+                try
+                {
+                    var clientTransport = _clientConnectionManager.CreateConnection(message, this);
+                    // Transfer channel and update client connection dictionary with real transport
+                    clientTransport.MessageChannel = transport.MessageChannel;
+                    _clientConnections.TryUpdate(connectionId, clientTransport, transport);
```

Comment:
>TryUpdate [](start = 39, length = 9)

this looks tricky

Reply:
Yes we shouldn't have to do this. The original feedback was to use a single dictionary entry instead of 2. We can just store a holder class with 2 fields, the channel and the transport.